File: 2305-fair-distribution-of-cookies/2305-fair-distribution-of-cookies.cpp
class Solution {
public:
    int ans = 8000000;
    int k1;
    vector<int> arr, ass;
    
    void backtrack(int index){
        if (index >= arr.size()){
            ans = min(ans, *max_element(ass.begin(), ass.end()));
            return;
        }
            

        if (ans <= *max_element(ass.begin(), ass.end()))
            return;

        for(int i=0; i < k1; i++){
            ass[i] += arr[index];
            backtrack(index+1);
            ass[i] -= arr[index];
        }
    }

    
    int distributeCookies(vector<int>& cookies, int k) {
        k1 = k;
        arr = cookies;
        sort(arr.rbegin(), arr.rend());
        ass.resize(k, 0);
        backtrack(0);
        
        return ans;
    }
};

File: 2305-fair-distribution-of-cookies/2305-fair-distribution-of-cookies.py
class Solution:
    def distributeCookies(self, cookies: List[int], k: int) -> int:
        
        if len(cookies) == k:
            return max(cookies)
         
        self.ans = float("inf")
        visited = set()
        ass = [0] * k
        
        self.mask = 0
        
        def backtrack(index):
            if (index, self.mask) in visited:
                return
            if index >= len(cookies):
                self.ans = min(self.ans, max(ass))
                return
            
            if self.ans <= max(ass):
                return
            
            for i in range(k):
                ass[i] += cookies[index]
                self.mask |= 1 << index
                visited.add((index, self.mask))
                backtrack(index+1)
                self.mask &= ~(1 << index)
                ass[i] -= cookies[index]
        
        backtrack(0)
        
        return self.ans